### PR TITLE
fix(compass-connections, sidebar): move all connection related logic and state to compass-connections COMPASS-8067

### PR DESCRIPTION
**Pull request by gribnoysup**
The goal of this patch is to fix various inconsistencies in connection logic caused by the logic being spread across two places in the application. The issue was highlighted by the autoconnection flow just not working as expected at all, but the fix was not possible to do in a clean way without refactoring the connection store as the store shape was just not accounting for multiple connection mode at all (it's honestly mostly luck that e2e tests were still passing, mostly because even in multiple connections mode we're still testing more or less a single connection flow at a time).

The main changes to the store that this patch introduces:

- Remove `activeConnectionId`, `connectingConnectionId` states: they don't make any sense in the multiple connection world
- Remove state that can be derived or already provided by other hooks, like connections list or connection status / status text
- Generally align the reducer following Redux style guide (which is only somewhat possible because React reducers work differently and doesn't allow to follow all the best practices): meaningful UI event based actions, no derived state in store, etc.
- Remove all the "legacy" prefixed methods and logic: there is nothing in the connection flow or connection state methods that would require us to export legacy methods and not share the same ones between two modes, all the switch logic can be done in place by just checking the feature flag if needed
- Move "Non genuine" modal and connection flow toasts to compass-connections: showing those is part of the connection flow logic and the UI is related to the connections

I want to note that it's not the final shape of this store, it's still not aligned with Compass architecture and using React reduxers here puts some restrictions on how separated the logic can be and the async nature of state updates spread around here.

For the reviewer: `connections-store.tsx` file is probably the main one to focus on, it changed a lot because very little of the old store was relevant and had to be adjusted for multiple connections to work, so it would probably be easier to just read through it as if it's a new file and not a diff.

I'll try to leave comments in places that I think are most notable, but if something doesn't make sense, please ask for clarifications!

### CODE REVIEW

File: packages/compass-connections/src/stores/connections-store.tsx [hunk 1]
```diff
@@ -289,160 +410,281 @@ export function useConnections({
...
+        });
+      };
+
+      const duplicate: ConnectionInfo = {
+        ...cloneDeep(connectionInfo),
+        id: new UUID().toString(),
+      };
+
+      if (duplicate.favorite?.name) {
+        const copyFormat = duplicate.favorite?.name.match(/(.*)\s\(([0-9])+\)/); // title (2) -> [title (2), title, 2]
```
Comment (addaleax) [inline packages/compass-connections/src/stores/connections-store.tsx:484]: Is this missing anchors? Otherwise we match something like `foo (2) bar` as well and lose everything after the closing parenthesis

```suggestion
        const copyFormat = duplicate.favorite?.name.match(/^(.*)\s\(([0-9])+\)$/); // title (2) -> [title (2), title, 2]
```

File: packages/compass-connections/src/stores/connections-store.tsx [hunk 2]
```diff
@@ -454,137 +696,91 @@ export function useConnections({
...
-      );
-      if (connection) {
         dispatch({
-          type: 'set-active-connection',
-          connectionInfo: connection,
+          type: 'connection-attempt-errored',
+          error: error as Error,
+          // Autoconnect flow might fail before we can even load connection info
+          // so connectionInfo might be undefiend at this point. In
+          // single-conneciton mode this requires some special handling so that
```
Comment (addaleax) [inline packages/compass-connections/src/stores/connections-store.tsx:717]:
```suggestion
          // single-connection mode this requires some special handling so that
```

**Review by addaleax — APPROVED** [review]

File: packages/compass-connections/src/components/legacy-connections.tsx [hunk 3]
```diff
@@ -76,6 +86,23 @@ const formCardLightThemeStyles = css({
   background: palette.white,
 });
 
+// Single connection form is a bit of a special case where form is always on
+// screen so even when user is not explicitly editing any connections, we need a
+// connection info object to be passed around. For that purposes this hook will
+// either return an editing connection info or will create a new one as a
+// fallback if nothing is actively being edited
+function useActiveConnectionInfo(
```
Comment (gribnoysup) [inline packages/compass-connections/src/components/legacy-connections.tsx:94]: It might be something that we maybe want to just make part of the connection-form logic, it's kinda convenient to have a default there and then we don't need to manually create new connection in two different places

File: packages/compass-connections/src/components/legacy-connections.tsx [hunk 4]
```diff
@@ -88,89 +115,69 @@ function Connections({
...
-      enableDebugUseCsfleSchemaMap,
-      protectConnectionStringsForNewConnections,
-    ]
+  const activeConnectionInfo = useActiveConnectionInfo(editingConnectionInfo);
+  const activeConnectionStatus = useConnectionInfoStatus(
+    activeConnectionInfo.id
   );
 
   const onConnectClick = (connectionInfo: ConnectionInfo) => {
-    void connect({ ...cloneDeep(connectionInfo) }, true).catch(() => {
```
Comment (gribnoysup) [inline packages/compass-connections/src/components/legacy-connections.tsx:153]: Honestly not sure why this was shallow and deep cloning, if someone knows what's the reason for this, please tell

File: packages/compass-connections/src/components/legacy-connections.tsx [hunk 5]
```diff
@@ -88,89 +115,69 @@ function Connections({
...
-    [
-      protectConnectionStrings,
-      forceConnectionOptions,
-      showKerberosPasswordField,
-      showOIDCDeviceAuthFlow,
-      enableOidc,
-      enableDebugUseCsfleSchemaMap,
-      protectConnectionStringsForNewConnections,
-    ]
+  const activeConnectionInfo = useActiveConnectionInfo(editingConnectionInfo);
```
Comment (gribnoysup) [inline packages/compass-connections/src/components/legacy-connections.tsx:138]: "Active" is like this concept only applicable to single connection form: what is being edited is also what you are connecting to (not sure if it's worth a comment in the code taking into account that we're going to remove it)

File: packages/compass-connections/src/components/non-genuine-connection-modal.tsx [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+const modalBodyStyles = css({
+  marginTop: spacing[3],
+  marginBottom: spacing[2],
+});
+
+export function showNonGenuineMongoDBWarningModal(
+  connectionInfo?: ConnectionInfo
+) {
+  return showConfirmation({
```
Comment (gribnoysup) [inline packages/compass-connections/src/components/non-genuine-connection-modal.tsx:24]: Button names are slightly different because of this change, but I'm not sure if it matters much tbh, I didn't try to make the names exactly the same

File: packages/compass-connections/src/components/non-genuine-connection-modal.tsx [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    title: 'Non-Genuine MongoDB Detected',
+    description: (
+      <>
+        <Banner variant={BannerVariant.Warning}>
+          {connectionInfo
+            ? `Server or service "${getConnectionTitle(connectionInfo)}"`
```
Comment (gribnoysup) [inline packages/compass-connections/src/components/non-genuine-connection-modal.tsx:30]: This modal wasn't taking multiple connections into account anymore, I changed the text a bit to be explicit about for which connection we show the modal

File: packages/compass-connections/src/connections-manager.ts [hunk 8]
```diff
@@ -14,6 +14,7 @@ import { cloneDeep, merge } from 'lodash';
 import { adjustConnectionOptionsBeforeConnect } from '@mongodb-js/connection-form';
 import mongodbBuildInfo from 'mongodb-build-info';
 import { openToast } from '@mongodb-js/compass-components';
+import { createCancelError, isCancelError } from '@mongodb-js/compass-utils';
```
Comment (gribnoysup) [inline packages/compass-connections/src/connections-manager.ts:17]: Drive-by, we have this helper we use elsewhere to create cancellation errors, using it here removes the need to checking error message to figure out if connection was canceled (which is not the most stable way of doing this)

File: packages/compass-connections/src/hooks/use-connection-repository.ts [hunk 9]
```diff
@@ -145,7 +168,7 @@ export function useConnectionRepository(): ConnectionRepository {
         infoToSave.connectionOptions?.connectionString
       );
 
-      if (!persistOIDCTokens) {
+      if (!preferences.getPreferences().persistOIDCTokens) {
```
Comment (gribnoysup) [inline packages/compass-connections/src/hooks/use-connection-repository.ts:171]: Drive-by to make the function more stable, we only care about the current value, we don't really need to update the whole callback for this, accessor works better here

File: packages/compass-connections/src/hooks/use-connection-repository.ts [hunk 10]
```diff
@@ -67,7 +81,16 @@ export function useConnectionRepository(): ConnectionRepository {
...
 
-  const persistOIDCTokens = usePreference('persistOIDCTokens');
+  const favoriteConnectionsRef = useRef(favoriteConnections);
+  favoriteConnectionsRef.current = favoriteConnections;
+
+  const nonFavoriteConnectionsRef = useRef(nonFavoriteConnections);
+  nonFavoriteConnectionsRef.current = nonFavoriteConnections;
+
+  const autoConnectInfoRef = useRef(autoConnectInfo);
+  autoConnectInfoRef.current = autoConnectInfo;
```
Comment (gribnoysup) [inline packages/compass-connections/src/hooks/use-connection-repository.ts:91]: Drive-by: in a lot of places we need to just access the current state of those, which without changing some of the methods here required subscribing to changes to those (invalidating dependencies of a loot of other hooks) even when we don't care about it and only want to read the current value

File: packages/compass-connections/src/stores/connections-store.tsx [hunk 11]
```diff
@@ -47,119 +42,238 @@ export function createNewConnectionInfo(): ConnectionInfo {
...
+// store and merging connection manager and connection repository state to this
+// store
 type State = {
-  activeConnectionId?: string;
-  activeConnectionInfo: ConnectionInfo;
-  connectingConnectionId: string | null;
-  connectingStatusText: string;
-  connectionErrorMessage: string | null;
-  oidcDeviceAuthVerificationUrl: string | null;
-  oidcDeviceAuthUserCode: string | null;
```
Comment (gribnoysup) [inline packages/compass-connections/src/stores/connections-store.tsx:57]: Literally none of this state makes sense in multiple connections mode 🙃

File: packages/compass-connections/src/stores/connections-store.tsx [hunk 12]
```diff
@@ -47,119 +42,238 @@ export function createNewConnectionInfo(): ConnectionInfo {
...
+          isEditingConnectionInfoModalOpen: action.autoDuplicate
+            ? // When autoduplicating we just save a copy right away without
+              // opening a editing modal
+              state.isEditingConnectionInfoModalOpen
+            : true,
+        };
+      case 'edit-connection': {
+        if (
+          // In multiple connections mode we don't allow to start another edit
+          // when one is already in progress
```
Comment (gribnoysup) [inline packages/compass-connections/src/stores/connections-store.tsx:196]: I'm not sure if this was an intentional decision, before this was not possible mostly because of the bugs in the state. Currently it should be technically possible to "open" multiple times while editing if you get a toast with an error for a different connection, but maybe it's still a good idea not to allow resetting the form while one is already opened

File: packages/compass-connections/src/stores/connections-store.tsx [hunk 13]
```diff
@@ -454,137 +696,91 @@ export function useConnections({
...
-        const copyFormat = duplicate.favorite?.name.match(/(.*)\s\(([0-9])+\)/); // title (2) -> [title (2), title, 2]
-        const name = copyFormat ? copyFormat[1] : duplicate.favorite?.name;
-        let copyNumber = copyFormat ? parseInt(copyFormat[2]) : 1;
-        while (findConnectionByFavoriteName(`${name} (${copyNumber})`)) {
-          copyNumber++;
-        }
-        duplicate.favorite.name = `${name} (${copyNumber})`;
+  const connectWithInflightCheck = useCallback(
+    async (connectionInfo: ConnectionInfo) => {
+      const inflightConnect = InFlightConnections.get(connectionInfo.id);
```
Comment (gribnoysup) [inline packages/compass-connections/src/stores/connections-store.tsx:757]: @himanshusinghs is this something that you had in mind?

File: packages/compass-sidebar/src/components/multiple-connections/sidebar.tsx [hunk 14]
```diff
@@ -486,33 +182,57 @@ export function MultipleConnectionSidebar({
...
-        />
-        <NonGenuineWarningModal
-          isVisible={genuineMongoDBModalVisible}
-          toggleIsVisible={setGenuineMongoDBModalVisible}
+          onOpenNonGenuineMongoDBModal={(connectionId: string) => {
+            showNonGenuineMongoDBWarningModal(connectionId);
+          }}
         />
+        {editingConnectionInfo && (
+          <ConnectionFormModal
```
Comment (gribnoysup) [inline packages/compass-sidebar/src/components/multiple-connections/sidebar.tsx:212]: Technically could've moved the form to compass-connections too already, but all the tests are also here and I'm trying to limit this already massive PR 😅

File: packages/compass-connections/src/stores/connections-store.tsx (hunk 1 above)
Comment (gribnoysup) [reply packages/compass-connections/src/stores/connections-store.tsx:484]: Good catch! I'll fix this implementation and add a test (we can also probably reduce the complexity of the algorithm there by reducing to the highest number right away instead of looping over for every number 🤔)

Comment (gribnoysup) [reply packages/compass-connections/src/stores/connections-store.tsx]: Updated in https://github.com/mongodb-js/compass/pull/6059/commits/ade3ada9978a3950ec2f0de12675b97b2252e08c, also fixed handling multi digit numbers

File: packages/compass-connections/src/stores/connections-store.tsx (hunk 13 above)
Comment (himanshusinghs) [reply packages/compass-connections/src/stores/connections-store.tsx:826]: yea something similar. I was probably thinking of having that in ConnectionsManager itself but having it here must have its own reasons. Sorry, I am yet to go through the PR (for my own understanding of the change).

Comment (gribnoysup) [reply packages/compass-connections/src/stores/connections-store.tsx:826]: My thinking is that this connect method is the main and only interface that other parts of UI should use to start a connection, so debounce should happen as high as possible. We are also planning to dissolve manager and repository in the connections redux store with the follow-up refactor, so this logic will end up in the store eventually anyway
